index.d.ts

```diff
@@ -0,0 +1,111 @@
+export interface Action<T> {
+  type: T;
+}
+
+
+/* reducers */
+
+export type Reducer<S> = <A extends Action<any>>(state: S, action: A) => S;
```

Comment:
As this is the only place the Action type is used, why is it generic?
Better make the type property any directly in the interface.